File: SWP_FRONTEND/frontend/src/utils/phoneValidator.js
// Phone Validator for Vietnamese mobile numbers
// Utility function to validate phone numbers in the frontend

// List of valid Vietnamese mobile prefixes
export const VIETNAMESE_MOBILE_PREFIXES = [
  // Viettel
  "086", "096", "097", "098", "032", "033", "034", "035", "036", "037", "038", "039",
  // Mobifone
  "070", "079", "077", "076", "078", "089",
  // Vinaphone
  "081", "082", "083", "084", "085", "088",
  // Vietnamobile
  "056", "058", "059",
  // Gmobile
  "052", "099"
];

/**
 * Validates a Vietnamese mobile phone number
 * - Must start with 0
 * - Must be 10 digits
 * - Must have a valid Vietnamese mobile prefix
 * 
 * @param {string} phone - The phone number to validate
 * @returns {boolean} - Whether the phone number is valid
 */
export const isValidVietnamesePhone = (phone) => {
  if (!phone || phone.trim() === '') {
    return false;
  }

  // Remove spaces and hyphens
  const cleanPhone = phone.replace(/\s+|-/g, '');

  // Check length and starting digit
  if (cleanPhone.length !== 10 || !cleanPhone.startsWith('0')) {
    return false;
  }

  // Check prefix
  const prefix = cleanPhone.substring(0, 3);
  return VIETNAMESE_MOBILE_PREFIXES.includes(prefix);
};

/**
 * Validates a phone number and returns an error message if invalid
 * 
 * @param {string} phone - The phone number to validate
 * @returns {string|null} - Error message or null if valid
 */
export const validatePhone = (phone) => {
  if (!phone || phone.trim() === '') {
    return 'Số điện thoại không được để trống';
  }

  // Remove spaces and hyphens
  const cleanPhone = phone.replace(/\s+|-/g, '');

  if (cleanPhone.length !== 10) {
    return 'Số điện thoại phải có đúng 10 chữ số';
  }

  if (!cleanPhone.startsWith('0')) {
    return 'Số điện thoại phải bắt đầu bằng số 0';
  }

  // Check prefix
  const prefix = cleanPhone.substring(0, 3);
  if (!VIETNAMESE_MOBILE_PREFIXES.includes(prefix)) {
    return 'Số điện thoại không đúng định dạng. Vui lòng nhập đúng đầu số điện thoại Việt Nam';
  }

  return null; // Phone is valid
};

/**
 * Format for displaying list of prefixes by provider
 * Useful for error messages or documentation
 */
export const formatPrefixesByProvider = () => {
  return `
    Viettel: 086, 096, 097, 098, 032, 033, 034, 035, 036, 037, 038, 039
    Mobifone: 070, 079, 077, 076, 078, 089
    Vinaphone: 081, 082, 083, 084, 085, 088
    Vietnamobile: 056, 058, 059
    Gmobile: 052, 099
  `;
};

File: SWP_BACKEND/SchoolMedicalManagementSystem/SchoolMedicalManagementSystem/src/main/java/group6/Swp391/Se1861/SchoolMedicalManagementSystem/util/PhoneValidator.java
package group6.Swp391.Se1861.SchoolMedicalManagementSystem.util;

/**
 * Tiện ích xác thực số điện thoại
 * Hỗ trợ kiểm tra và xác thực số điện thoại theo định dạng Việt Nam
 */
public class PhoneValidator {

    /**
     * Danh sách đầu số điện thoại di động của Việt Nam
     */
    private static final String[] VIETNAMESE_MOBILE_PREFIXES = {
            "086", "096", "097", "098", "032", "033", "034", "035", "036", "037", "038", "039", // Viettel
            "070", "079", "077", "076", "078", "089", // Mobifone
            "081", "082", "083", "084", "085", "088", // Vinaphone
            "056", "058", "059", // Vietnamobile
            "052", "099" // Gmobile
    };

    /**
     * Xác thực số điện thoại di động Việt Nam
     * - Phải bắt đầu bằng số 0
     * - Phải có 10 chữ số
     * - Phải thuộc các đầu số nhà mạng Việt Nam
     *
     * @param phone Số điện thoại cần kiểm tra
     * @return true nếu số điện thoại hợp lệ, false nếu không hợp lệ
     */
    public static boolean isValidVietnamesePhone(String phone) {
        if (phone == null || phone.isEmpty()) {
            return false;
        }

        // Loại bỏ khoảng trắng và dấu gạch ngang nếu có
        phone = phone.replaceAll("\\s+|-", "");

        // Kiểm tra độ dài và bắt đầu bằng số 0
        if (phone.length() != 10 || !phone.startsWith("0")) {
            return false;
        }

        // Kiểm tra đầu số
        String prefix = phone.substring(0, 3);
        for (String validPrefix : VIETNAMESE_MOBILE_PREFIXES) {
            if (prefix.equals(validPrefix)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Kiểm tra và trả về thông báo lỗi nếu số điện thoại không hợp lệ
     *
     * @param phone Số điện thoại cần kiểm tra
     * @return Thông báo lỗi hoặc null nếu hợp lệ
     */
    public static String validatePhone(String phone) {
        if (phone == null || phone.isEmpty()) {
            return "Số điện thoại không được để trống";
        }

        // Loại bỏ khoảng trắng và dấu gạch ngang nếu có
        String cleanPhone = phone.replaceAll("\\s+|-", "");

        if (cleanPhone.length() != 10) {
            return "Số điện thoại phải có đúng 10 chữ số";
        }

        if (!cleanPhone.startsWith("0")) {
            return "Số điện thoại phải bắt đầu bằng số 0";
        }

        // Kiểm tra đầu số
        String prefix = cleanPhone.substring(0, 3);
        boolean validPrefix = false;
        for (String validPre : VIETNAMESE_MOBILE_PREFIXES) {
            if (prefix.equals(validPre)) {
                validPrefix = true;
                break;
            }
        }

        if (!validPrefix) {
            return "Số điện thoại không đúng định dạng. Vui lòng nhập đúng đầu số điện thoại Việt Nam";
        }

        return null; // Số điện thoại hợp lệ
    }
}
